dll/win32/ws2_32/src/startup.c

```diff
@@ -74,6 +102,22 @@ WSACleanup(VOID)
         /* Decrement process reference count and check if it's zero */
         if (!(RefCount = InterlockedDecrement(&Process->RefCount)))
         {
+            /* Be sure no other thread is using WSAThread */
+            WsCreateWSAThreadSynchronization();
```

Comment:
Thank you for your reply. Most people only replied for code conventions errors I made (or told me they wouldn't reply at all anymore because I made to many) and I was starting to feel really unappreciated here. So thanks for starting this substantive discussion :).

I understand this can be dangerous, it can hang the application. The reason why I put this here is because I thought that if an other thread is still using the WSA API it could crash the application if it already starts cleaning up resources. My question is, is this a possible scenario (because if not for some reason this can be deleted right on the spot). If this a possible scenario, will have to choose between the risk of hanging the application or let it crash in some scenarios.

Reply:
Sorry about that, we do not want you to feel unappreciated.
We have moved just recently to github (well, a few months) so it is a bit new to everyone, and we are still learning.


As for your code, I think you have creating and using a critical section mixed up:
You probably want `EnterCriticalSection` here, and `LeaveCriticalSection` in `WsDestroyWSAThreadSynchronization`.

I'd suggest to create the critical section in `DLL_PROCESS_ATTACH`, with `InitializeCriticalSection(&pWSAThreadLock);`,
and to clean it up in `DLL_PROCESS_DETACH` with `DeleteCriticalSection(&pWSAThreadLock);`